plugins/usbdmx/ShowJockey.cpp

```diff
@@ -0,0 +1,393 @@
+      if (n_left <= 0) {
+        break;
+      }
+
+      ret_val = m_adaptor->BulkTransfer(handle, endpoint, buffer, n_packet,
+                                       &transferred, URB_TIMEOUT_MS);
+
+      if (ret_val) {
+        break;
+      }
+
+      buf += transferred;
+      n_left -= transferred;
+      n_packet = max_packet_size;
+      transferredAll += transferred;
+    }
+
+    if (transferredAll == size) {
+      return transferredAll;
+    }
+  }
+
+  return -1;
+}
+
+// SynchronousShowJockey
+// -----------------------------------------------------------------------------
+
+SynchronousShowJockey::SynchronousShowJockey(
+    LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    const std::string &serial)
+    : ShowJockey(adaptor, usb_device, serial) {
+}
+
+bool SynchronousShowJockey::Init() {
+  libusb_device_handle *usb_handle;
+
+  int interface_number;
+  int endpoint;
+  int max_packet_size;
+  if (!LocateInterface(m_adaptor,
+                       m_usb_device,
+                       &interface_number,
+                       &endpoint,
+                       &max_packet_size)) {
+    return false;
+  }
+
+  bool ok = m_adaptor->OpenDeviceAndClaimInterface(m_usb_device,
+                                                   interface_number,
+                                                   &usb_handle);
+  if (!ok) {
+    return false;
+  }
+
+  std::auto_ptr<ShowJockeyThreadedSender> sender(
+      new ShowJockeyThreadedSender(m_adaptor,
+                                   m_usb_device,
+                                   usb_handle,
+                                   max_packet_size,
+                                   endpoint));
+  if (!sender->Start()) {
+    return false;
+  }
+
+  m_sender.reset(sender.release());
+  return true;
+}
+
+bool SynchronousShowJockey::SendDMX(const DmxBuffer &buffer) {
+  return m_sender.get() ? m_sender->SendDMX(buffer) : false;
+}
+
+// ShowJockeyAsyncUsbSender
+// -----------------------------------------------------------------------------
+class ShowJockeyAsyncUsbSender : public AsyncUsbSender {
+ public:
+  ShowJockeyAsyncUsbSender(LibUsbAdaptor *adaptor,
+                           libusb_device *usb_device,
+                           int endpoint,
+                           int max_packet_size_out,
+                           libusb_device_handle *handle)
+                           : AsyncUsbSender(adaptor, usb_device),
+                             m_endpoint(endpoint),
+                             m_max_packet_size_out(max_packet_size_out) {
+    m_usb_handle = handle;
+    m_tx_frame = NULL;
+  }
+
+  ~ShowJockeyAsyncUsbSender() {
+    CancelTransfer();
+    delete[] m_tx_frame;
+  }
+
+  libusb_device_handle* SetupHandle() {
+    return m_usb_handle;
+  }
+
+  bool PerformTransfer(const DmxBuffer &buffer) {
+    uint16_t nb_sequence = DMX_MAX_SLOT_NUMBER / (m_max_packet_size_out - 2);
+    int final_size = DMX_MAX_SLOT_NUMBER + 2 * nb_sequence;
```

Comment:
Can we have some brackets round this please. I assume you mean DMX_MAX_SLOT_NUMBER + (2 * nb_sequence)?